<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>An introduction to the general linear model</title>
    <meta charset="utf-8" />
    <meta name="author" content="Alex Sanchez-Pla" />
    <meta name="date" content="2023-05-06" />
    <script src="libs/header-attrs-2.21/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/metropolis.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/metropolis-fonts.css" rel="stylesheet" />
    <script src="libs/kePrint-0.0.1/kePrint.js"></script>
    <link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
    <link rel="stylesheet" href="mycss.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# An introduction to the general linear model
]
.subtitle[
## Application to biomarker discovery with omics data.
]
.author[
### Alex Sanchez-Pla
]
.institute[
### Departamento de Genética, Microbiología y Estadística (UB) <br> Unidad de Estadística y Bioinformática (VHIR)
]
.date[
### 2023-05-06
]

---






# Outline

.columnwide[
  ### 1) [The general linear models](#review)
  ### 2) [Some examples of simple linear models](#examples)
  ### 3) [Linear models with R](#lm)
  ### 4) [Linear models for omics data](#limma)
  ### 5) [Exercises](#exercises)
]

---

class: inverse, middle, center

name: review

# A recap on linear models &lt;a id="review"&gt;&lt;/a&gt;

---

# What is a linear model (in statistics)

- Linear models appear when we assume that the relation between:

  - one variable, 
  - and another (set of) variable(s)
  
  can be represented through a linear relation, that is, one of the form
  $$
  y \simeq a + b\times x \quad \mbox{ or:} \quad y \simeq a + b\times x + c \times z + \dots
  $$
- Linear models have been thoroughly used because of their flexibility and many of the important techniques in statistics are "just" linear models.

---

# Regression or  ANOVA are linear models

- Linear models provide a flexible way to modelling relations and building explanatory or predictive models, as seen in **linear regression**.

$$
Y\_i = \beta\_0 + \beta\_1 \times X\_{1i} + \beta\_2 \times X\_{2i}  + \dots+ \epsilon\_i
$$

- But they also provide a convenient setting to 
      - **describe** experimental designs and 
      - to **analyze data** that has been obtained from experiments performed according to the design described, as seen in **ANOVA**.
    
$$
Y\_{ij}=\mu\_i+e\_{ij}=\mu+\tau\_i+e\_{ij}, \quad i=1\dots k,\quad j=1\dots r.
$$

---

# The General Linear model

- This capability of linear models to embrace *apparently distinct* models leads to introduce the notion o General Linear Models*.

- GLMs constitute a versatile statistical framework used to analyze relationships between *dependent* (or *response* or *predicor*) variables and one or more independent (or *explanatory* or *predictive*) variables. 
- A GLM can be written as:

`$$Y_i = \beta_0 + \beta_1 X_{i1} + \beta_2 X_{i2} + \ldots + \beta_p X_{ip} + \epsilon_i$$`

  - where, if there are `\(i=1,\dots,n\)` observations, 
    - `\(Y_i\)` represents the dependent variable for the `\(i\)`th observation, 
    - `\(X_{ij}\)` represents the `\(j\)`th independent variable for the `\(i\)`th observation, 
    - `\(\beta_0\)` represents the intercept, 
    - `\(\beta_1, \beta_2, \ldots, \beta_p\)` represent the coefficients for each independent variable, and 
    - `\(\epsilon_i\)` represents the error term for the `\(i\)`th observation. 

---


# Why is the GLM so "General"? 

- The *generality* of the GLM comes from the fact that it can be used to describe, and analyze in a somehow unified form, a variety of problems.

- Although one may think that, assuming that a relation is linear, is an excessive simplification, this approach
  - is thoroughly used,
  - is very successful.
  
- Indeed, it can be shown that many common statistical models can be re-written as linear models:

    - [Common statistical tests are linear models](https://lindeloev.github.io/tests-as-linear/)


---

# The GLM has multiple extensions

- Apart of its extensive use "as is",
- The general linear model can be extended in multiple directions
  - Generalized linear models
  - Mixed linear and non linear models
  - etc.
  
---

# Matrix form for the general linear model
  
- The equation expressing the linear relationship between the dependent variable and multiple independent variables  can be written in a *compact form*, that is, for all individuals at once, using matrix notation as:

$$
Y = X\beta + \epsilon
$$
  - This notation will be useful later when we discuss linear models for omics data analysis.

---

class: inverse, middle, center

name: review

# Some examples of linear models &lt;a id="examples"&gt;&lt;/a&gt;


---

# t-test and ANOVA as GLMs

- Consider a study comparing __two diets__ in mice, a "standard" vs a "high-fat" one.

- The main goal of the study are:

  - To *estimate* the mean weight of mice after having been nurrished with each diet
  - To *compare* the difference in weight between the standard and "high fat" diets.
  
- The usual approach for this problem can be:  
  - consider two variables, representing the weight of mice nurrished by ech diet 
      - `\(X_1\sim N(\mu_1, \sigma_1)\)`, `\(X_2\sim N(\mu_2, \sigma_2)\)`, or, without assuming normality
      - `\(X_1\)`, `\(X_2\)` belong to some continuous distribution.
    - And test the hypothesis 
      `\(H_0: \mu_1=\mu_2\)` with a `\(t-test\)` or a non parametric one, such as Wilcoxon's test.
  
---

# Re-write the problem as a linear model

- An alternative approach may come by re-writing the problem as a linear model.
- If _we consider that the weight of the animals is a linear function of the diets_, the following linear model can be written:

$$ 
Y_i = \beta_0 + \beta_1 x_i + \varepsilon_i, i=1,\dots,N 
$$
- where:
  - `\(Y_i\)` represents the weights of the i-th experimental unit-
      &lt;!-- - The term *experimental unit* describes the `\(N\)` different entities from which we obtain a measurement. In this case this is "mice". --&gt;
  - `\(x_i = 1\)` when mouse `\(i\)` receives the high fat diet and &lt;br&gt;  `\(x_i = 0\)`  when it receives the standard diet.
      &lt;!-- - We call thees variables *indicator variables* since they simply indicate if the experimental unit had a certain characteristic or not. --&gt;
      



---

# Comparison of 2 groups (t-test)

- The coefficients represent the (differential) effect of diet
- Comparison between diets can be based on the means or on the coefficients.

.center[ 
&lt;img align = "center" src="images/linMod1.png" width = "90%"&gt; 
&lt;p&gt;
&lt;img align = "center" src="images/linMod2.png" width = "80%"&gt; 
 ]


---

# Comparison of three groups (ANOVA)

.pull-left[
.center[ 
&lt;img align = "center" src="images/linMod5.png" width = "100%"&gt; 
&lt;p&gt;
&lt;img align = "center" src="images/linMod4.png" width = "100%"&gt; 
 ]
]
.pull-right[
- With more than two diets the scenario is similar
  - `\(\beta_0\)` represents the "control" or standard
  - `\(\beta_1\)` , `\(\beta_2\)` the effect of each diet added to the standard
- Comparisons can be based on
  - The means
  - The coefficients
]

- Notice that, although the linear model postulates a relation between _parameters_ we will work with their _estimates_ so, there is an error that we need to control in order for our conclusions (inferences) to be reliable.
---

# Estimating a linear model

- For linear models to be useful, we have to _estimate_ the unknown values `\(\beta_i\)`.

- The standard approach in science is to find the values that minimize the distance of the fitted model to the data. 

- The following is called the least squares (LS) equation:

&lt;!-- $$ --&gt;
&lt;!-- RSS = \sum_{i=1}^n \left\{  y_i - \left(\beta_0 + \beta_1 x_i \right)\right\}^2 --&gt;
&lt;!-- $$ --&gt;

.center[ 
&lt;img align = "center" src="images/linMod7RSS.png" width = "50%"&gt;
]
- This quantity is called the residual sum of squares (RSS). 

- Once we find the values that minimize the RSS, we will call the values the _least squares estimates (LSE)_ and denote them with  `\(\hat \beta_i\)`.


---

# Matrix notation for linear models

- Linear models can be written as:
  - explicit expressions (above) or 
  - using matrix notation (below)

.center[ 
&lt;img align = "center" src="images/linMod8MatrixNotation.png" width = "90%"&gt;
]

- Matrix notation is generally preferred.
  - More compact notation
  - More efficient computations.
  
---

&lt;!-- # Computers prefer matrices --&gt;

&lt;!-- - Many languages are ready to perform matrix operations in parallel which yields a smaller number of operations and a smaller computing time  --&gt;
&lt;!-- - Higher efficency is also due to the implicit use of matrix algebra libraries. --&gt;

&lt;!-- &lt;p&gt; --&gt;

&lt;!-- ![](images/captura_007.png) --&gt;

&lt;!-- --- --&gt;

# Two groups in matrix notation

&lt;p&gt;
.center[
![](images/captura_008.png)
]

This _linear model_ can be written in more compact notations as:

$$
\mathbf{Y}=\mathbf{X}\boldsymbol{\beta}+\boldsymbol{\varepsilon}
$$
The matrix `\(\mathbf{X}\)` is called the _design matrix_
---

# Three groups in matrix notation

.center[
![](images/captura_009.png)
]

---

# Fitting linear models

- A linear model can be _fitted_ by solving the _normal equations_

![](images/captura_010.png)

- Error estimates for the model coefficients can also be obtained:

![](images/captura_012.png)

---

# Significance testing with linear models.

- Assuming a series of assumptions hold

  - Variance homogeneity
  - Linearity of relations
  - Independence and _normality_ of error terms
  
- A test can be built to test the significance of the model coefficients


![](images/captura_011.png)


---

# More linear models: crossed designs 

![](images/captura_014.png) 

- We assume effects are additive!

---
# Crossed designs with interaction 

![](images/captura_015.png)

- Interaction shows additional to additive effects 

---

# How is a linear models usually applied?

- Estimating the parameters values to know mean effect of one factor or of each level od this factor.
  
- Comparing the parameters' values  to decide if different treatments have the same effect, or have any effect at all.
  
- Adjusting the effect of distinct covariables, which may not be of direct interest ion the studyu but whose effect may interfere in the relation between the response variable and the explanatory ones.
  
  - A common quote in papers: *"The study of ... controlled for age, race, and sex ..."*,
  - Usually, what it means is that we have included thes variables in the model expecting the variability that they explain is separated from that explained buy the other components of the model.
        ![](images/captura_016.png)


---

class: inverse, middle, center

name: review

# Linear models in R &lt;a id="lm"&gt;&lt;/a&gt;


---

# Fitting linear models in R

- In practice, when using R, _we rarely fit a model by solving the normal equations_

- The usual, and most practical way to do it, is to use `lm` function.

- The `lm` function requires

  - Either a `formula` relating the variables to be included in a linear model
  
  - Or a design matrix, that we can create using the `model.matrix` function, and which, implicitly, defines that model.

---

# The importance of the design matrix

- The choice of design matrix is a _critical step_ in linear modeling as

  - it encodes which coefficients will be fit in the model, 
  - and the inter-relationship between the samples.

--

- Defining which design matrix we use is equivalent to defining the parameters of the model (or the model's _parametrization_).

  - Same data can be modelled differently, if parameters receive different meanings.

  - In practice this represents using distinct design matrices
  
- A typical example: How does the meaning of a one-way factor ANOVA change if we consider a model with or without an intercept?
  - How is it reflected in the design matrix?

---

# Model matrix for two groups

- Suppose we have two groups, 1 and 2, with two samples each.
- We might start to encode this experimental design like so:


```r
x &lt;- c(1,1,2,2)
f &lt;- formula(~ x)
model.matrix(f)
```

```
##   (Intercept) x
## 1           1 1
## 2           1 1
## 3           1 2
## 4           1 2
## attr(,"assign")
## [1] 0 1
```

---

# Model matrix for two groups (2)

- Note that an intercept will be included by default, so the formula could equivalently be written: `~ x + 1`.

- We can then inspect the design matrix which is formed by this:


```r
model.matrix(f)
```

```
##   (Intercept) x
## 1           1 1
## 2           1 1
## 3           1 2
## 4           1 2
## attr(,"assign")
## [1] 0 1
```

---

# model.matrix requires factors

.pull-left[
- Note, this is not the design matrix we wanted.
- We should instead first tell R that these values should not be interpreted numerically, but as different levels of a factor variable:

]

.pull-right[


```r
x &lt;- factor(c(1,1,2,2))
model.matrix(~ x)
```

```
##   (Intercept) x2
## 1           1  0
## 2           1  0
## 3           1  1
## 4           1  1
## attr(,"assign")
## [1] 0 1
## attr(,"contrasts")
## attr(,"contrasts")$x
## [1] "contr.treatment"
```
]

- Now we have achieved the correct design matrix.
- Or have we?
---

# The role of the intercept term

.pull-left[
- Note that the previous matrix has one intercept column and one group column although there are two groups indeed.

  - The first group's values are represented by the basal or "overall mean".
  - The second group's are represented by one column.

- An alternative representation is possible setting the intercept to zero.

- Both representations are equivalent and for one-factor designs it's up to you which one to choose
]

.pull-right[

```r
x &lt;- factor(c(1,1,2,2))
model.matrix(~ x + 0)
```

```
##   x1 x2
## 1  1  0
## 2  1  0
## 3  0  1
## 4  0  1
## attr(,"assign")
## [1] 1 1
## attr(,"contrasts")
## attr(,"contrasts")$x
## [1] "contr.treatment"
```
]


---

# Design matrix for more than 2 groups

- How is the design matrix for an experiment with 3 groups?
- We proceed like in the previous case


```r
x &lt;- factor(c(1,1,2,2,3,3))
model.matrix(~ x)
```

```
##   (Intercept) x2 x3
## 1           1  0  0
## 2           1  0  0
## 3           1  1  0
## 4           1  1  0
## 5           1  0  1
## 6           1  0  1
## attr(,"assign")
## [1] 0 1 1
## attr(,"contrasts")
## attr(,"contrasts")$x
## [1] "contr.treatment"
```

- Again the first group is implicit in the intercept but it can be set explicitly by setting the intercept to zero.

---

# An alternative parametrization

-   An alternate formulation of design matrix is possible by specifying `+0` in the formula:


```r
x &lt;- factor(c(1,1,2,2,3,3))
model.matrix(~ x + 0)
```

```
##   x1 x2 x3
## 1  1  0  0
## 2  1  0  0
## 3  0  1  0
## 4  0  1  0
## 5  0  0  1
## 6  0  0  1
## attr(,"assign")
## [1] 1 1 1
## attr(,"contrasts")
## attr(,"contrasts")$x
## [1] "contr.treatment"
```

- This representation allows fitting a separate coefficient for each group.

---


class: inverse, middle, center

name: limma

# Linear Models for Omics Data &lt;a id="limma"&gt;&lt;/a&gt;

---

# Linear models for omics data

- For the analysis of omics data a very popular option is the `limma` package.

- `limma` extends some R functionalities to make them easy to use in the analysis of omics data using linear models.

- Besides this it includes extensions to the standard linear model to improve analysis capabilities.

- In the following we show

  - How to create a design matrix from a "targets" file containing information on groups.
  - How to create a contrasts matrix to define the comparisons to be done.
    -   How to do the comparisons and how to interpret the resulting analysis tables.

---

# Example: Comparing 3 types of tumors

- This example study is based on a paper published in  http://www.ncbi.nlm.nih.gov/pubmed/15897907 whose data are available in GEO as series GSE1561 series on the following link
http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE1561

- The researchers investigated three types of breast cancer tumors: apocrine (APO), basal (BAS) and luminal (LUMI). 

- The classification is based on the resistance of tumors to estrogen and androgen receptors.

  - Tumors classified as "APO" are negative for estrogen receptor (ER-) and positive for the androgen receptor (AR+).
  - Those classified as "LUMI" are ER + and AR + and
  - Those classified as "BAS" are ER- and AR.

---

# Identifying groups and comparisons

- The assignment of each sample to an experimental group can be obtained from this link:
http://www.ncbi.nlm.nih.gov/geo/gds/profileGraph.cgi?gds=1329

- Obviously this is an observational study but its analysis can be done using a linear model approach as well.

- We will usually proceed in three steps:

1. Identify the experimental factors and their levels.
2. Write the design matrix associated with this study design.
3. Build the contrast matrix that defines the comparisons we are interested in.

- In this example we have identified three groups and we wish to compare each tumor type with the oher two

    1. "APO" vs “LUMI”
    2. “APO" vs “BAS”
    3. “LUMI" vas "BAS"

---

# Defining the groups

- An easy way to define the study groups consists of preparing a text file where these are established.


```r
url &lt;- "https://raw.githubusercontent.com/ASPteaching"
repo &lt;- "Introduction_to_Design_of_Experiments"
folder &lt;- "main/omicsData/dataset_2_Breast_cancer_GSE1561/data"
dataFile &lt;- "BreastCancerGSE1561.csv"
targetsFile &lt;- "targets.txt"
targetsFileName &lt;- paste(url, repo, folder, targetsFile, sep="/")
```

---
# The `targets`file


```r
library(magrittr)
targets &lt;-read.table(targetsFileName, row.names=1, head=T)
```

&lt;small&gt;

```r
kableExtra::kable(head(targets[,1:6]), n=7) %&gt;% kableExtra::kable_styling()
```

&lt;table class="table" style="margin-left: auto; margin-right: auto;"&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt;   &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; fileName &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Sample &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Ids &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; SampleIDs &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Group &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Apocrine.grade &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; GSM26878.CEL &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; GSM26878 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF14 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; EnPnT2N1G2 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF14 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 3 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; GSM26883.CEL &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; GSM26883 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF19 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; EnPuT4N0Gu &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF19 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 2 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; GSM26887.CEL &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; GSM26887 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF23 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; EnPnT2N0G2 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF23 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 3 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; GSM26903.CEL &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; GSM26903 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF39 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; EnPuT4N0Gu &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF39 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 3 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; GSM26910.CEL &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; GSM26910 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF46 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; EnPnT4N1G3 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF46 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 3 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; GSM26888.CEL &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; GSM26888 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF24 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; EnPnTiN0G3 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PF24 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; B &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 2 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;
_Showing only first 7 rows_
&lt;/small&gt;
---

# Creating the design matrix

.pull-left[

```r
design&lt;-matrix(
  c(1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,1,1,1,1,1,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,1,1,1,1,1),
  nrow=15,byrow=F)
colnames(design) &lt;-c("A", "B", "L")
rownames(design)&lt;-  targets$Sample
```
]

.pull-right[

```r
print(design)
```

```
##      A B L
## PF14 1 0 0
## PF19 1 0 0
## PF23 1 0 0
## PF39 1 0 0
## PF46 1 0 0
## PF24 0 1 0
## PF25 0 1 0
## PF28 0 1 0
## PF34 0 1 0
## PF42 0 1 0
## PF15 0 0 1
## PF32 0 0 1
## PF33 0 0 1
## PF43 0 0 1
## PF47 0 0 1
```
]
---

# Another way to create the design matrix 

.pull-left[
&lt;small&gt;

```r
design2 &lt;-model.matrix(~ 0+targets$Group)
colnames(design2)&lt;-c("A", "B", "L")
rownames(design2)&lt;-  targets$Sample
```
&lt;/small&gt;
]

.pull-right[

```r
print(design2)
```

```
##      A B L
## PF14 1 0 0
## PF19 1 0 0
## PF23 1 0 0
## PF39 1 0 0
## PF46 1 0 0
## PF24 0 1 0
## PF25 0 1 0
## PF28 0 1 0
## PF34 0 1 0
## PF42 0 1 0
## PF15 0 0 1
## PF32 0 0 1
## PF33 0 0 1
## PF43 0 0 1
## PF47 0 0 1
## attr(,"assign")
## [1] 1 1 1
## attr(,"contrasts")
## attr(,"contrasts")$`targets$Group`
## [1] "contr.treatment"
```
]
---

# Defining the questions: The contrast matrix


```r
library(limma)
cont.matrix &lt;- makeContrasts (
  AvsB = B-A,
  AvsL = L-A,
  BvsL = L-B,
  levels=design)
cont.matrix
```

```
##       Contrasts
## Levels AvsB AvsL BvsL
##      A   -1   -1    0
##      B    1    0   -1
##      L    0    1    1
```

---

# Data for the analysis

- Data can be obtained from the open repository Gene Expression Omnibus but, for simplicity, it has been downloaded and prepared for the analysis.

- They are also available remotely at the sub-directory "omicsData"


```r
dataFile &lt;- "BreastCancerGSE1561.csv"
dataFileName &lt;- paste(url, repo, folder, dataFile, sep="/")

dataMatrix &lt;- read.csv(dataFileName, row.names=1)

colnames(dataMatrix)==rownames(targets)
```

```
##  [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
## [11] TRUE TRUE TRUE TRUE TRUE
```

```r
dim(dataMatrix)
```

```
## [1] 22283    15
```

---

# Fit the model and the contrasts

- Once we have the data, the design matrix and the contrast matrix we can proceed to estimate the model,  fit the contrasts and check the results


```r
fit&lt;-lmFit(dataMatrix, design)
fit.main&lt;-contrasts.fit(fit, cont.matrix)
fit.main&lt;-eBayes(fit.main)
```

- This creates an object, which is called here `fit.main` from where distinct results can be extracted.

---

# Results are in "topTables"

- For each comparison in the contrast matrix a "top Table" can be generated showing features sorted from most to least differentially expressed, based on the test p-value.


```r
topTab_AvsB &lt;- topTable (fit.main, number=nrow(fit.main), coef="AvsB", adjust="fdr")
kableExtra::kable(head(topTab_AvsB, n=5)) %&gt;% kableExtra::kable_styling()
```

&lt;table class="table" style="margin-left: auto; margin-right: auto;"&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt;   &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; logFC &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; AveExpr &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; t &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; P.Value &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; adj.P.Val &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; B &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 204667_at &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; -3.038344 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 8.651157 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; -17.65663 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1.40e-06 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 13.610350 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 215729_s_at &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 3.452290 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 6.137595 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 13.67095 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 2.08e-05 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 11.214963 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 202579_x_at &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1.293975 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 10.226164 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 12.26570 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 5.24e-05 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 10.115252 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 209787_s_at &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 2.023113 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 9.871298 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 12.05888 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 5.24e-05 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 9.939472 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 214243_s_at &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; -5.015130 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 9.278980 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; -11.47539 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 6.75e-05 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 9.422507 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

```r
# topTab_AvsL &lt;- topTable (fit.main, number=nrow(fit.main), coef="AvsL", adjust="fdr"); head(topTab_AvsL)
# topTab_BvsL  &lt;- topTable (fit.main, number=nrow(fit.main) , coef="BvsL", adjust="fdr"); head(topTab_BvsL)
```

---

# Volcano plots provide visualization

.pull-left[
- A volcano plot shows, for each comparison
  - the magnitude of the change ("biological significance") vs
  - the "statistical significance" (-log p-value)
]

.pull-right[
.center[

```r
volcanoplot(fit.main, coef="AvsB", 
            highlight=10)
```

&lt;img src="An-introduction-to-the-General-Linear-Model_files/figure-html/showResults-1.png" width="100%" /&gt;
]
]
---

class: inverse, middle, center

name: exercises

# Exercises &lt;a id="exercises"&gt;&lt;/a&gt;

---

# Exercises on linear models for omics

- You can go through the exercises in the document [Exercises on linear models for microarrays](https://github.com/ASPteaching/Introduction_to_Design_of_Experiments/blob/56a927d76a67378ca08985864fc998df35a61c43/Exercices%20in%20linear%20models%20and%20experimental%20design.pdf)

  - Start by reading the experiment description
  - Postulate the type of experimental design
  - Write the linear model
  - Create the design and the contrast matrix
  - Get the data
  - Fit the model 
  - Examine the results
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
